lib/elixir/lib/module/types/infer.ex

```diff
@@ -102,26 +104,84 @@ defmodule Module.Types.Infer do
 
   # %{...}
   def of_pattern({:%{}, _meta, args} = expr, context) do
-    result =
-      expr_stack(expr, context, fn context ->
-        map_reduce_ok(args, context, fn {key, value}, context ->
-          with {:ok, key_type, context} <- of_pattern(key, context),
-               {:ok, value_type, context} <- of_pattern(value, context),
-               do: {:ok, {key_type, value_type}, context}
-        end)
+    case expr_stack(expr, context, &of_pairs(args, &1)) do
+      {:ok, pairs, context} -> {:ok, {:map, pairs_to_unions(pairs, context)}, context}
+      {:error, reason} -> {:error, reason}
+    end
+  end
+
+  # %var{...}
+  def of_pattern({:%, _meta1, [var, {:%{}, _meta2, args}]} = expr, context) when is_var(var) do
+    expr_stack(expr, context, fn context ->
+      with {:ok, pairs, context} <- of_pairs(args, context),
+           {var_type, context} = new_var(var, context),
+           {:ok, _, context} <- unify(var_type, :atom, context) do
+        pairs = [{{:literal, :__struct__}, var_type} | pairs]
+        {:ok, {:map, pairs_to_unions(pairs, context)}, context}
+      end
+    end)
+  end
+
+  # %Struct{...}
+  def of_pattern({:%, _meta1, [module, {:%{}, _meta2, args}]} = expr, context)
+      when is_atom(module) do
+    struct_pairs =
+      Enum.map(:maps.remove(:__struct__, module.__struct__()), fn {key, value} ->
+        {term_to_type(key), term_to_type(value)}
```

Comment:
I am not sure if we can assume `term_to_type` for the value. The default value can be `nil` for example or have an empty list but in practice be any value. The best assumption at this point is "dynamic".

Reply:
Ah yes, of course. Maybe we shouldn't include the fields from the struct at all since this matches: `%File.Stat{} = %{__struct__: File.Stat}`? Alternatively, if we introduce a struct type like you mentioned below this may not be an issue.

Reply:
Yes, exactly. My initial reaction is that treating them strictly would be better.